File: Data Structures & Algorithms/combination-target-sum-ii/submission-28.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) { 
        const result = []  
        candidates.sort()

        function dfs(start, total, cur) {
            if (total === target) {
                result.push([...cur]) 
                return
            } 

            if (start >= candidates.length || total > target) {
                return
            } 

            for (let i = start; i < candidates.length; i ++) { 
                if (i > start && candidates[i] === candidates[i - 1]) continue
                cur.push(candidates[i])
                dfs(i + 1, total + candidates[i], cur) 
                cur.pop()
            }
        } 

        dfs(0, 0, []) 

        return result
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-28.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = [] 

        def dfs(start, cur, part): 
            if cur < 0: 
                return 

            if cur == 0: 
                res.append(part.copy()) 
                return 

            for i in range(start, len(nums)): 
                rest = cur - nums[i] 
                part.append(nums[i]) 
                dfs(i, rest, part) 
                part.pop() 
        dfs(0, target, [])
        return res
